Refactoring of FuzzyQuery

alexksikes opened the pull request:
Relates to #10217

This PR is against the query-refactoring branch.

MaineC commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java:
```diff
@@ -19,40 +19,31 @@
 
 /**
  *
  */
```
Can you please add some text or remove the empty comment?

MaineC commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Override
+    protected Query createExpectedQuery(FuzzyQueryBuilder queryBuilder, QueryParseContext context) throws IOException {
```
Much like with the tests for SimpleQueryStringQuery this method is almost a copy of the original toQuery code in the QueryBuilder. Can we track the reasoning behind this type of testing here: https://github.com/elastic/dev/blob/master/design/queries/general-guidelines.md (I believe we've discussed it in multiple PR threads and face2face but have lost of all the specific arguments).

javanna replied:
I will take care of this.

MaineC commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -55,49 +74,81 @@
+     * Sets the boost for this query.  Documents matching this query will (in addition to the normal
+     * weightings) have their score multiplied by the boost provided.
+     */
+    @Override
```
I think we can simply re-use the comment of the super method by leaving this one undocumented?

MaineC commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -19,35 +19,54 @@
     public static final String NAME = "fuzzy";
 
-    private final String name;
+    public static final Fuzziness DEFAULT_FUZZINESS = Fuzziness.AUTO;
```
Personally I like having at least a brief line of JavaDoc on anything that is public that explains it's purpose...

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -140,4 +185,94 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+
+    static MultiTermQuery.RewriteMethod parseRewriteString(QueryParseContext parseContext, @Nullable String rewrite) {
+        MultiTermQuery.RewriteMethod rewriteMethod = null;
+        if (parseContext.isFilter()) {
```
I think the filter flag needs to be checked earlier, already in the parser. `isFilter` is a statefull flag in the QueryParseContext that gets set/unset when parsing one of the (former) filters. ParseContext.isFilter at this point doesn't reflect the state it was in during traversing the query tree. Check QueryParseContext.parseInnerFilterToQueryBuilder() to see where state changes can occur.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,145 @@
+        }
+        expectedQuery.setBoost(queryBuilder.boost());
+
+        if (queryBuilder.queryName() != null) {
```
I think we shouldn't add the expectedQuery to the namedQuery in the context here, that should be done in toQuery() and it's part of what's beeing tested later.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -19,35 +19,54 @@
+    private final String fieldName;
 
-    private Integer maxExpansions;
+    private final Object text;
```
If this stays an Object we have a problem with builder equality when we do the roundtrip doXContent->fromXContent that we currently do in the tests. I had similar problems in TermQueryBuilder already. Imagine an Integer value her (like its used in the current test setup), that is rendered to JSON (or other XContent) via `writeValue(Object value)` which inspects the class of the Object and then chooses a type-specific method to write it. On the parser side, this is always read back as a String with `parser.text()`. So if we check those builders for equality we will get `126 != "126"`. For the lucene query produced in the end this won't matter, but I think it would be great to keep that property that when we have builder A, generate XContent for that and parse it to Builder B, they should be the same.
One option here would be to push up the conversion to String we do in `toQuery()` anyway to the constructor already. Also I'm not sure if the java API currently is to lenient here by allowing any `Object`, since only objects that correctly parse back as Strings will really work with the current parser code. Another option might be to allow parsing to numeric values using `objectText()` like in CommonTermsQueryParser.

cbuescher commented:
Did a round of review, agree with most of the changes. I left some comments, mostly concerning questions of where to set default Rewrite method when in filter context and some thoughts concerning leaving internal query value representation as `Object`. We might have to check if it is possible to restrict the Java-API to only accept Numeric, String and Date values (at least that's what seems allowed reading the docs). Then we might be able to just have one internal String representation.

alexksikes commented:
Thanks for the review. I went for storing `value` as a String internally. I'm entirely sure this is the right decision though. We should probably only accept Numeric, String and Date values.

cbuescher commented:
@alexksikes I went through your last commit, left a few comments. Sorry I didn't realize they show up in such a strange way if I do the comments on the commit directly here in Github, will try to avoid this in the future.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -140,4 +185,82 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(this.fieldName);
+        out.writeGenericValue(this.value);
```
Since `value` internally is a String now, we can change read/writer here as well.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -55,49 +78,77 @@
-    public FuzzyQueryBuilder(String name, Object value) {
-        this.name = name;
-        this.value = value;
+    public FuzzyQueryBuilder(String fieldName, Object value) {
```
If we already break the builder API by changing the value-getter, it would be great to somehow also restrict the constructor to not accept every value type anymore. Maybe deprecate this constructor and instroduce new ones for String and Number? Not sure if also Date type is handled correctly when parsed to String here. 
In general I'm unsure here to how far this PR should change the status quo. I think only selected Object types worked here in the past, it should be stated somehere (docs?) what actually works. In our own code we really only use the constructor with String argument, so at least internally we could easily change that, but not sure how many client code this would break.

javanna replied:
don't we need constructors for other supported type slike we do in term query?

javanna replied:
sorry didn't see them, can you move them up please?

alexksikes replied:
sure

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java:
```diff
@@ -63,15 +51,12 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
-        MultiTermQuery.RewriteMethod rewriteMethod = null;
-        if (parseContext.isFilter()) {
-            rewriteMethod = MultiTermQuery.CONSTANT_SCORE_REWRITE;
-        }
```
From the discussion before I think that removing the special case for FuzzyQuery inside filter context is okay here since CONSTANT_SCORE_REWRITE ist the default MultiTermQuery rewrite method anyway. 
@jpountz could you have a look at this change since I think it's related to the filter/query merging.

jpountz replied:
CONSTANT_SCORE_REWRITE is the rewrite for all multi-term queries but FuzzyQuery: see the FuzzyQuery constructor which calls `setRewriteMethod(new MultiTermQuery.TopTermsBlendedFreqScoringRewrite(maxExpansions));`

cbuescher replied:
Thanks, seems that unfortunately we need to re-set it then when we're in filter context. @alexksikes I guess that means we have to go back to using the String constants from `QueryParser#parseRewriteMethod`. I was wondering if it makes sense to factor out the constants used there into a separate enum or a fixed list of constants to make the dependency of the Parser to those constants more explicit. I see that the second part of `QueryParser#parseRewriteMethod` also does some suffix-magix with the values passed in, but at least for the constant we use in FuzzyQueryParser it would be good to link it somehow to the code there. Or whats your take on this?

alexksikes replied:
OK we should only rewrite if we have a filter or if `rewrite` is explicitly set, otherwise the default rewrite on  FuzzyQuery is untouched. If we want to make the `rewrite` options as constants, we would have to do so for every query that supports `rewrite`. Not sure this is really necessary.

cbuescher commented:
Went through the changes another time, would like for @javanna to have another look and for an opinion about how much we should restrict the current constructor that allows (but won't work) for any `Object` argument

alexksikes commented:
Thanks for the review. I restricted the set of accepted values for now.

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -252,4 +264,40 @@ private int termLen(String text) {
+        } else if (value instanceof Float) {
+            return new Fuzziness((Float) value);
+        } else if (value instanceof String) {
+            return new Fuzziness((String) value);
```
maybe add a new constructor Fuzziness(Object) instead of these ifs?

javanna replied:
@alexksikes did you see this comment?

alexksikes replied:
Not sure why it is rendered like this in github, but I am using a factory Fuzziness.newFuziness.

javanna replied:
I see but I wonder if we should remove this method and simply accept Object and add a Fuzziness constructor that accepts Object instead

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -106,9 +115,12 @@ public static Fuzziness parse(XContentParser parser) throws IOException {
                     }
                 } catch (NumberFormatException ex) {
-                    return build(fuzziness);
+                    try {
```
can you explain the reason behind this change?

alexksikes replied:
This is to properly parse floats into floats and not into strings, so that the query from the parser is the same as from the builder. I would have errors such as "0.6" != 0.6.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -19,35 +19,59 @@
  * A Query that does fuzzy matching for a specific value.
  */
-public class FuzzyQueryBuilder extends MultiTermQueryBuilder implements BoostableQueryBuilder<FuzzyQueryBuilder> {
+public class FuzzyQueryBuilder extends MultiTermQueryBuilder<FuzzyQueryBuilder> implements BoostableQueryBuilder<FuzzyQueryBuilder> {
```
implements BoostableQueryBuilder should be gone by now

javanna commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -19,35 +19,59 @@
 
     private String rewrite;
 
+    private float boost = 1.0f;
```
boost will be gone once you rebase

javanna commented:
I had a quick look at this and left a few comments. There is an inconsistency between parser and builder here in the existing code. The builder supports an object for the value mamber but the parser only parses a string. I am not 100% sure which of the two is wrong. Feels like fuzzy query should allow for an object similar to what the term query does, but then our MappedFieldType#fuzzyQuery method takes a string as argument, so object wouldn't really work out, that would lead us to support strings only then? Thoughts @jpountz ?

javanna commented:
Also @alexksikes would be nice if you can rebase, it is hard to review at this point after #11974 got in. Thanks!

jpountz commented:
@javanna I looked at how classes that extend MappedFieldType implement fuzzyQuery, and they seem to all need to perform some parsing of the string value. So maybe we could make it an object and update all fuzzyQuery implementations to parse the object to the appropriate type (Date, Long, ...) using exactly the same logic as the termQuery method?

javanna commented:
Agreed @jpountz thanks for looking. We will do that upstream rather than in the query-refactoring branch then. Makes sense @alexksikes ?

jpountz commented:
+1 to doing it in master

alexksikes commented:
Yes agreed. Thanks @jpountz.

alexksikes commented:
@javanna @cbuescher Thanks for the review. I rebased the PR and updated it. I had to add a fix which will be removed in order to avoid NPE on unmapped fields in test.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -31,18 +42,31 @@
+
+    /** Default as to whether transpositions should be treated as a primitive edit operation, 
+     * instead of classic Levenshtein algorithm. Defaults to false. */
+    public final static boolean DEFAULT_TRANSPOSITIONS = false;
```
nitpick: can you change these constants to static final rather than final static (just a common convention on order of modifiers)

alexksikes replied:
ha right, didn't see this, thanks!

javanna commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -31,18 +42,31 @@
 
     //LUCENE 4 UPGRADE  we need a testcase for this + documentation
-    private Boolean transpositions;
+    private Boolean transpositions = DEFAULT_TRANSPOSITIONS;
```
wonder if we should make these Integer and Boolean just int and boolean primite types.

javanna replied:
this would make sense especially given that their setters accept primitive types

alexksikes replied:
right makes sense

javanna commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryParser.java:
```diff
@@ -19,59 +19,50 @@
+        String rewrite = null;
         if (parseContext.isFilter()) {
-            rewriteMethod = MultiTermQuery.CONSTANT_SCORE_REWRITE;
+            rewrite = "constant_score";
```
why don't we have a constant for this anymore?

alexksikes replied:
that's a constant for RewriteMethod class?

javanna replied:
I dont follow can you clarify?

alexksikes replied:
Sorry I might be missing something. rewrite is a String and rewriteMethod is a RewriteMethod object. So we can't assign MultiTermQuery.CONSTANT_SCORE_REWRITE to rewrite, unless we make rewrite a RewriteMethod.

javanna replied:
you are not missing anything, it is me, now I get it. maybe we can use one of the existing constants mentioned in the other comments then?

alexksikes replied:
Ha that yes, sounds perfect!

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -25,6 +25,7 @@
 import org.elasticsearch.cluster.ClusterService;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.ParseFieldMatcher;
```
is this import needed?

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,132 @@
+public class FuzzyQueryBuilderTest extends BaseQueryTestCase<FuzzyQueryBuilder> {
+
+    private static final String[] REWRITE_METHOD = new String[]{"constant_score", "scoring_boolean", "constant_score_boolean", 
+            "top_terms_1", "top_terms_boost_1", "top_terms_blended_freqs_1"};
```
I think we have constants for this now: https://github.com/elastic/elasticsearch/blob/8238f497d85d785a61ebc017b8ac96fc5fcd28ae/core/src/main/java/org/elasticsearch/index/query/support/QueryParsers.java#L32

alexksikes replied:
I was going to use them but they are private

javanna replied:
it makes sense to make them public them I think

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,132 @@
+            "top_terms_1", "top_terms_boost_1", "top_terms_blended_freqs_1"};
+
+    private static final String[] MAPPED_FIELD_NAME = new String[] { INT_FIELD_NAME, DOUBLE_FIELD_NAME, DATE_FIELD_NAME, 
+            STRING_FIELD_NAME };
```
don't we have mappedFieldNames already in BaseQueryTestCase?

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,132 @@
+            }
+            query = new FuzzyQueryBuilder(fieldName, text);
+        } else {
+            query = new FuzzyQueryBuilder(randomAsciiOfLengthBetween(1, 10), randomAsciiOfLengthBetween(1, 50));
```
I think it would be important to test unmapped fields against other types besides String. Thats what BaseTermQueryTestCase does in doCreateTestQueryBuilder. I think we should do exactly the same here no?

javanna replied:
this comment wasn't addressed I think

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,132 @@
+        }
+        if (expectedQuery instanceof MultiTermQuery) {
+            // NO COMMIT: fix to be removed to avoid NPE on unmapped fields
+            context.parseFieldMatcher(randomBoolean() ? ParseFieldMatcher.EMPTY : ParseFieldMatcher.STRICT);
```
I see why this happens, will fix s we can remove this hack

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,132 @@
+        assertThat(fuzzyQueryBuilder.validate().validationErrors().size(), is(1));
+
+        fuzzyQueryBuilder = new FuzzyQueryBuilder("field", null);
+        assertThat(fuzzyQueryBuilder.validate().validationErrors().size(), is(1));
```
can we test also the case where we might get two errors?

alexksikes replied:
Yes I added the check as well.

javanna commented:
I did another round of review

alexksikes commented:
@javanna Thanks for the review. I updated the PR accordingly.

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -22,43 +22,56 @@
 
+    /** the prototype constant is intended for deserialization when used with 
+     * {@link org.elasticsearch.common.io.stream.StreamableReader#readFrom(StreamInput)} */
+    public static final Fuzziness PROTOTYPE = AUTO;
```
maybe have a public static readFrom method here that can be called from FuzzyQueryBuilder readFrom so that the prototype can become package private?

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -22,43 +22,56 @@
+        this((Object) fuzziness);
+    }
+    
+    public Fuzziness(Object fuzziness) {
```
this should not be public, private is fine

alexksikes replied:
indeed

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -22,43 +22,56 @@
     private Fuzziness(float fuzziness) {
+        this((Object) fuzziness);
         Preconditions.checkArgument(fuzziness >= 0.0 && fuzziness < 1.0f, "Valid similarities must be in the interval [0..1] but was [" + fuzziness + "]");
-        this.fuzziness = fuzziness;
```
I think I would revert this change, no need to call the other constructor here, so the precondition can happen first, which makes more sense. Same in other constructors

alexksikes replied:
yes

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,151 @@
+
+    @Override
+    protected FuzzyQueryBuilder doCreateTestQueryBuilder() {
+        // unmaped fields
```
s/unmaped/unmapped

javanna commented:
left a few more minor comments

alexksikes commented:
Thanks for the review. I updated the PR.

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,151 @@
+                    value = randomBoolean();
+                    break;
+            } // OBJECT_FIELD_NAME and STRING_FIELD_NAME assigned to random string
+        }
```
I left this comment before but it didn't get addressed yet. I think it would be important to test unmapped fields against other types besides String. Thats what BaseTermQueryTestCase does in doCreateTestQueryBuilder. I think we should do exactly the same here.

alexksikes replied:
I created a new helper method getRandomFieldNameAndValue which I moved up to BaseQueryTestCase.

javanna commented:
left one comment around testing, LGTM otherwise

alexksikes commented:
@javanna I updated the PR. Parsing Fuzziness is tricky as it is type dependant. On String field, fuzziness is an edit distance which can be a String with "AUTO", or an int with value 0, 1, 2 or floats in [0-1). For other types, fuzziness translates into a range query and is kept as a String if it does not fall within the given values above.

alexksikes commented:
@javanna I updated the PR. Parsing Fuzziness is tricky as it is type dependant. On String field, fuzziness is an edit distance which can be a String with "AUTO", or an int with value 0, 1, 2 or floats in [0-1). For other types, fuzziness translates into a range query and is kept as a String if it does not fall within the given values above.

alexksikes commented:
@javanna I rebased the PR and took into account #12229. There seems to be a bug with "AUTO" for which I'll submit an issue.

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -187,19 +193,24 @@ public byte asByte() {
 
+    // A deserialized fuzziness with AUTO is not the static Fuzziness.AUTO object
+    private boolean isAuto() {
+        return this == AUTO || fuzziness.equalsIgnoreCase("AUTO");
```
can't you simply use equals here? do we need ignoreCase?

alexksikes replied:
We do if we want to be consistent with the previous behavior. "AUTO" was parsed case insensitively.

javanna replied:
then I guess the equals method is wrong too cause auto != AUTO . I think I would try to uppercase (or lowercase) when parsing and make this case insensitive that way.

alexksikes replied:
Here the equals means that the current Object is actually the AUTO Object which happens when parsing in Fuzziness#parse. So either we have the AUTO Object or we have a new Object that has a fuzziness with "AUTO".

javanna replied:
Maybe my previous comment wasn't clear. I think that if we want to make this class case-insensitive, our current equals impl is wrong too, because "auto" is not equal to "AUTO" but isAuto would return true. I think we should clean this up. Can we simply uppercase the fuzziness string when we parse?

alexksikes replied:
Sure we could clean this up a bit by uppercasing "auto" in the builder, and then remove all checks with .equalsIgnoreCase. Sounds good?

javanna replied:
sounds good

cbuescher replied:
Did this change in the builder or does the case check stay? Can't find it, maybe missed it.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -174,4 +217,72 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        }
+        if (query == null) {
+            int maxEdits = fuzziness.asDistance(BytesRefs.toString(value));
+            query = new FuzzyQuery(new Term(fieldName, BytesRefs.toBytesRef(value)), maxEdits, prefixLength, maxExpansions, transpositions);
```
do we need this `BytesRefs.toBytesRef(value)` conversion?

alexksikes replied:
This comes from this PR https://github.com/elastic/elasticsearch/pull/12020/files#diff-8a2008f5cb78b0eecefab5ef414d7231R124. This is using the same logic as termQuery in the way it handles Object.

javanna replied:
that looks like it's not needed though, isn't the value always a BytesRef once parsed?

alexksikes replied:
it is parsed with XContentParser#objectBytes which returns a value of type Object.

javanna replied:
ok if it uses the same logic as termQuery why don't we need any conversion there? I think that we always convert to bytes ref in the constructor internally anyway, so we can rely on the fact that value is BytesRef?

alexksikes replied:
I think it is only converted internally to a BytesRef if it is a String?

javanna replied:
yes and that should be enough.

alexksikes replied:
which we do if the field is mapped. Otherwise we need to create a new Term which takes a BytesRef. I think this is more of a best of effort kinda scheme.

javanna replied:
gotcha, I didn't realize that we were creating a new query not going through the mapper, which accepts a Term. This is correct as-is, sorry.

javanna commented on core/src/test/java/org/elasticsearch/common/unit/FuzzinessTests.java:
```diff
@@ -162,4 +164,29 @@ public void testAsDistance() {
+        assertEquals(fuzziness.asInt(), deserializedFuzziness.asInt());
+    }
+
+    private Fuzziness doSerializeRoundtrip(Fuzziness in) throws IOException {
```
this could be static

alexksikes replied:
sure will change that, thanks.

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -60,23 +62,26 @@
+    protected static final String DATE_FIELD_NAME = "mapped_date";
+    protected static final String OBJECT_FIELD_NAME = "mapped_object";
+    protected static final String[] mappedFieldNames = new String[] { STRING_FIELD_NAME, INT_FIELD_NAME,
+            DOUBLE_FIELD_NAME, BOOLEAN_FIELD_NAME, DATE_FIELD_NAME, OBJECT_FIELD_NAME };
```
what changed here? I can't spot the difference compared to before

alexksikes replied:
just the ordering the change, which seems more logical, first String, then Int, then Double, Boolean, Date and finally Object.

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -132,13 +137,13 @@ protected void configure() {
                     DOUBLE_FIELD_NAME, "type=double",
                     BOOLEAN_FIELD_NAME, "type=boolean",
-                    STRING_FIELD_NAME, "type=string",
+                    DATE_FIELD_NAME, "type=date",
```
can you explain the reason behind this change?

alexksikes replied:
just the ordering changed

javanna commented on core/src/test/java/org/elasticsearch/test/ElasticsearchTestCase.java:
```diff
@@ -383,6 +383,11 @@ public static String randomRealisticUnicodeOfCodepointLength(int codePoints) {
+    
+    public static String randomTimeValue() {
+        final String[] values = new String[]{"d", "H", "ms", "s", "S", "w"};
+        return randomIntBetween(1, 3) + randomFrom(values);
```
why between 1 and 3 ?

alexksikes replied:
sure this could be 0d, 0H etc .. will change

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -382,6 +387,65 @@ protected static String getRandomRewriteMethod() {
+    /**
+     * Helper method to return a random field (mapped or unmapped) and a value
+     */
+    protected static Tuple<String, Object> getRandomFieldNameAndValue() {
```
I think that this method duplicates logic that's already existing in other tests. If so can we use it in other tests too?

alexksikes replied:
yes but it would be better to make this change in another PR?

javanna replied:
potentially but I don't see why we add a method in the base class if we don't use it everywhere. then add it to the class that needs it only?

alexksikes replied:
I can move it back to FuzzyQueryBuilderTest. But note that many queries could make use of this starting with Term Query.

javanna replied:
then let's use it now otherwise this change is half baked

alexksikes replied:
wouldn't it be better to make the change only local to FuzzyQuery, and then have another PR in which we make use of this method for all the other queries? In which case I'm happy to move it back to FuzzyQueryBuilderTest.

javanna replied:
I think it's a simple enough change to make as part of this PR.

alexksikes replied:
ok we settled for doing it in another PR, thanks @javanna

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -382,6 +387,65 @@ protected static String getRandomRewriteMethod() {
+        return fuzziness;
+    }
+
+    protected static boolean isNumericFieldName(String fieldName) {
```
to be honest I think we have enough static helper methods in the test base class, can we simply check the name of the field in the fuzzy test that needs this?

alexksikes replied:
this is more for readability, also there are other types which are numerical which in the future we might want to check against. I can still make that change if you'd like though.

javanna replied:
ok lets keep it but please let's not create a list for this, we can just do if INT.equals(field) || DOUBLE.equals(field)

alexksikes replied:
sure thanks, that seems better.

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -382,6 +387,65 @@ protected static String getRandomRewriteMethod() {
+        return new Tuple(fieldName, value);
+    }
+    
+    protected static Fuzziness randomFuzziness(String fieldName) {
```
what's the reason for this to be in the test base class? is it going to be used in other tests?

alexksikes replied:
A couple of other queries have a fuzziness option, so it will be handy to have them share this part of the code.

javanna commented:
left come comments

alexksikes commented:
@javanna I addressed the comments and rebased. Thanks for the review.

javanna commented on core/src/test/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -383,6 +383,11 @@ public static String randomRealisticUnicodeOfCodepointLength(int codePoints) {
+    
+    public static String randomTimeValue() {
+        final String[] values = new String[]{"d", "H", "ms", "s", "S", "w"};
+        return randomIntBetween(0, 3) + randomFrom(values);
```
didn't we say that we would change this?

alexksikes replied:
I thought you meant between 0 - 3?

javanna replied:
no I meant that I don't understand why we randomize between 0 and 3. Can you elaborate?

alexksikes replied:
yes I guess that is more of a fantasy related to the name of the method, although 0s or 0d could be corner cases.

javanna replied:
Lemme clarify I mean that if we add this method to ESTestCase it should be generic enough. I guess there are much more valid values possible for a time value?

alexksikes replied:
OK then the value will be 0 - 1000

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTest.java:
```diff
@@ -0,0 +1,98 @@
+        context.setAllowUnmappedFields(true);
+        
+        FuzzyQueryBuilder fuzzyQueryBuilder = new FuzzyQueryBuilder(STRING_FIELD_NAME, "text");
+        fuzzyQueryBuilder.fuzziness(Fuzziness.build(randomFrom("a string which is not auto", "3h", "200s")));
```
can we not validate this situation before toQuery? where does the exception get thrown at the moment exactly?

alexksikes replied:
Those are field type dependant. For example, a string field can't use fuzziness with a time distance. I could check for a fuzziness string which is not "AUTO" though.

javanna replied:
where is the NumberFormatException thrown?

alexksikes replied:
this happens because the fieldType is not of the correct type when performing this fuzzy query. For string fields, we can only have fuzziness be an edit distance, not a time distance nor a String other than AUTO. So the asInt() method of Fuzziness is called and calls as a last resort Float.parseFloat (meaning maybe the user tried to type 1.0) which then returns a NumberFormatException.

javanna replied:
ok thanks for the explanation. I guess if we can detect this earlier and add some intelligence to the validate method?

alexksikes replied:
i'd think we would need the fieldType in validate for that

javanna replied:
right, maybe then add some clearer validation in toQuery given the filedType ? I don't love this NumberFormatException, would love to have clearer error messages.

javanna replied:
we agreed that this is not that easy to do as it depends on mappings, we will leave as-is for now

javanna commented:
I did another round and replied to your comments. @cbuescher wanna have a quick look too?

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -174,4 +217,72 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        FuzzyQueryBuilder fuzzyQueryBuilder = new FuzzyQueryBuilder(in.readString(), in.readGenericValue());
+        fuzzyQueryBuilder.fuzziness = Fuzziness.readFuzzinessFrom(in);
+        fuzzyQueryBuilder.prefixLength = in.readVInt();
+        fuzzyQueryBuilder.maxExpansions = in.readVInt();
```
I think those two fields are now non-optional, so could use readInt()?

alexksikes replied:
since these are more likely to be small, better use the `V` methods. There are always written and yes there are none optional.

cbuescher replied:
Oh, I'm sorry I mixed VInt with optional, my fault. Must be the heat. Please ignore.

alexksikes replied:
No worries, the heat does affect me too :-)

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -174,4 +217,72 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        out.writeGenericValue(this.value);
+        this.fuzziness.writeTo(out);
+        out.writeVInt(this.prefixLength);
+        out.writeVInt(this.maxExpansions);
```
same as above

cbuescher replied:
sorry, my bad.

cbuescher commented:
@alexksikes did a very quick round and left only a few comments, mostly small stuff and questions for clarfication

alexksikes commented:
@javanna rebased and comments addressed. Thanks for the review as always.

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -45,13 +47,20 @@
 
     private Fuzziness(String fuzziness) {
-        this.fuzziness = fuzziness;
+        if (fuzziness == null) {
```
this could be replaced with Objects.requireNotNull

alexksikes replied:
i see it throws an NPE exception, not as neat as the error we return to the user, unless you want to wrap the call in a try catch and still return the nicer error to the user.

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -187,19 +196,24 @@ public byte asByte() {
     }
 
+    // A deserialized fuzziness with AUTO is not the static Fuzziness.AUTO object
+    private boolean isAuto() {
```
can't this method go away now?can't we just do AUTO.equals(this) ?

alexksikes replied:
sure but isAuto() reads better no?

javanna commented:
left a small comment

javanna commented on core/src/main/java/org/elasticsearch/common/unit/Fuzziness.java:
```diff
@@ -121,7 +130,7 @@ public int asDistance() {
 
     public int asDistance(String text) {
-        if (this == AUTO) { //AUTO
+        if (fuzziness.equals("AUTO")) { //AUTO
```
I meant to leverage our own equals method rather than a string comparison. AUTO.equals(this) ?

alexksikes replied:
ha yes!

javanna commented:
LGTM
